Add check of hashed data when writing new data

peytondmurray opened the pull request:
This PR adds a check when writing data to a file. This check verifies that chunks of data that already exist in the hashtable correspond to arrays that match the data we are trying to write.

This should incur a performance overhead when writing datasets that reuse chunks, but ensures we won't be corrupting hashtables in the future. Closes #294.


Note: Currently I'm seeing a test failure locally with `test_delete_string_dataset` on `master`. The failure happens when trying to open the file after running `h5repack` on the `hdf5` file, with `OSError: Unable to open file (bad object header version number)`. I'm opening this PR to see if CI sees it.

ArvidJB commented:
The same check should also go into `write_dataset_chunks`.

Can you add a test where the hash function is mocked to be one of the previous data_version hash functions? The check should get triggered.

peytondmurray commented:
Looks like there must be a problem locally with `h5repack` that is responsible for the failing test I'm seeing, and anyway it isn't a failing test related to this change.

---

## Changes

- Removed a little dead code
- Changed the logic slightly to make it more readable by removing the usage of `setdefault`. This let me get the clarity I needed to figure out how to carry out the check of the hashed data.
- Added check of the hashed data
- Moved some code inside the branch statement that only needed to be run on one branch

peytondmurray commented:
Okay, I think we have our bases covered with these tests.

ArvidJB commented on versioned_hdf5/backend.py:
```diff
@@ -124,23 +127,143 @@ def write_dataset(f, name, data, chunks=None, dtype=None, compression=None,
+    if len(arr) > 0 and isinstance(arr[0], bytes):
+        return arr
+    else:
+        return np.vectorize(lambda i: bytes(i, encoding='utf-8'))(arr)
```
You can use `np.char.encode` here, I think?

peytondmurray replied:
Normally I definitely would but this doesn't work for object dtype arrays:

![image](https://github.com/deshaw/versioned-hdf5/assets/14017872/cb7b3c73-95f5-4cb8-aff4-5ffd6d4c268e)

ArvidJB reviewed (approved):
Looks good to me!
